Account for a microinverter connected at the generator input of the deye hp3 inverter

Johnny1206 opened the pull request:
Added register 667 to total pv power to account for a microinverter connected at the generator input of the deye inverter.

Fix https://github.com/evcc-io/evcc/issues/19171, fix https://github.com/evcc-io/evcc/issues/20235

Johnny1206 committed:
Update deye-hybrid-hp3.yaml

Added register 667 to total pv power to account for a microinverter connected at the generator input of the deye inverter.

premultiply commented:
Unsure if we should add this by default?

Usage as diesel generator input should not be handled as solar production.

Johnny1206 committed:
Consider usage of the generator port

Added an advanced parameter to consider the usage of the generator port. By default the power on the generator port won't be added to the total pv power now.

Johnny1206 commented:
Added an advanced parameter to consider the usage of the generator port. Choices are derived from the generator port usage options in the manual of the inverter. 
By default the power on the generator port won't be added to the total pv power now.
Code has not been tested!

andig commented on templates/definition/meter/deye-hybrid-hp3.yaml:
```diff
@@ -30,6 +30,10 @@ params:
     advanced: true
   - name: maxacpower
+  - name: GenPortUse
+    choice: ["Generator", "SmartLoad", "MicroInverter"]
```
I don't understand the options. Wouldn't "inlcude generator port" be better? Why doesn't this affect energy?

Johnny1206 replied:
yes, you could absolutely change this to a Boolean for its current usage. 

My approach was to make it expandable for more settings in the future. The three options are taken from the inverter manual. One could say, if the GEN-Port is used in “Generator” mode it may be useful for example to add it to the grid power. Or in “smart load” it is more related to house consumption. And like I already implemented: In “MicroInverter” Mode, it should be added to PV power.

How it can be added to total pv energy however, I do not know. I haven’t found a register related to total generator port energy.

I only really need the Micro-Inverter feature and therefore I would be happy if I had an option for the power on the generator port to be added to the Total PV power in one way or another.

andig replied:
@Johnny1206 will energy then be wrong since its only pv?

Johnny1206 committed:
Fix datatype of register 667

The datatype of the generator port power is signed because the power at this port can flow in both directions.

Johnny1206 commented:
I have noticed that there is also a high word for the generator power with register address 671 but as my micro inverter power does not exceed 32 kW, I can not test the implementation. 
The correct implementation is probably to also add register 671 and scale it with a factor of 2^16 but i am unsure.

turbotorsten commented:
If this change gets approved, would the changes also work for the deye-hybrid-3p? As far as I noticed the main difference seems to be the scaling of the values?

premultiply commented:
@Johnny1206 Could you please attach the current register documentation this is based on? Thanks!

Johnny1206 commented:
At the moment i use the Deye.MODBUS.RTU.V104.pdf from dy-support.org ([3497-Deye-MODBUS-RTU-V104.pdf](https://github.com/user-attachments/files/19176367/3497-Deye-MODBUS-RTU-V104.pdf)). Unfortunately a lot of information in the document is in chinese, so searching the document via STRG+F is not always helpful.


My current understanding is the following:

- Power from the Generator Port: Low Word in 667 and High Word in 671 as a signed 32 bit integer with a scaling of 1 W
- Total Energy from the Generator Port: Low Word in 537 and High Word in 538 also as a signed 32 bit integer with a scaling of 0.1 kWh.
- Register 133 indicates if the Port is configured as generator, smart load or micro inverter. This could potentialy be used instead of the configuration parameter i suggested.

I am unsure on how to add together register 667 and 671. My guess would be to scale 671 by 2^16 and add it to 667 but i do not know if this works for signed values.

Johnny1206 commented:
@premultiply 
could you please give an update on whether you still plan to implement this?

andig commented:
Ping @premultiply wie wäre es den downstream Port optional zu integrieren? Unabhängig von seinem Zweck, das würde ich dem Anwender überlassen? 

In jedem Fall sollten wir hie rzu einem Ergebnis kommen.

andig commented:
@Johnny1206 mangels Feedback: willst du PR so umbauen dass wir eine einfache "include generator port" Option bauen?

Johnny1206 committed:
revised implementation of gen port option

- changed parameter to boolean "includegenport"
- added energy registers of gen port to the template

Johnny1206 commented:
@andig: habs wie vorgeschlagen abgeändert. allerdings kann ichs nicht testen. Daher bitte mal akribisch reviewen, ob das funktionieren kann, was ich da gebastelt hab. 
Außerdem weiß ich nicht, wie ich das High-Word der Leistung am Generatorport da mit rein krieg. Deshalb funktioniert die Implementierung jetzt nur bis 32kW am Gen-Port.

andig commented:
>  allerdings kann ichs nicht teste

Was fehlt Dir?

andig commented:
> ußerdem weiß ich nicht, wie ich das High-Word der Leistung am Generatorport da mit rein krieg.

32 statt 16 bit auslesen?

Johnny1206 commented:
@andig 

> Was fehlt Dir?
ich kriegs nicht gebaut bei mir.

> 32 statt 16 bit auslesen?
so hab ichs bei der energie gemacht. aber bei der leistung sind die addressen nicht aufeinanderfolgenden: 667 und 671. Da weiß ich nicht weiter.

andig commented:
ah, lass uns das ignorieren

> ich kriegs nicht gebaut bei mir.

Musst du nicht: https://github.com/evcc-io/evcc/pull/19495

Johnny1206 commented:
@andig
danke für den tipp. cooles feature. so kann ichs easy testen.

jetzt muss ich nur noch rausfinden warum der bei "{{- if eq .includegenport true }}" das bool nicht frisst und "incompatible types for comparison" meldet.

Johnny1206 committed:
fixed incompatible datatype issue

put the "true" in the if statement in quotes:
{{- if eq .includegenport "true" }}

Johnny1206 commented:
@andig 
So, Problem gefixt -> siehe commit.

Hab das Template jetzt auch durch einbinden der yaml testen können: Das funktioniert soweit. Die Leistung des Balkonkraftwerks am GEN-Port wird bei includegenport=true jetzt zur PV-Leistung addiert und bei includegenport=false weiter wie vorher auch nicht berücksichtigt und somit vom Hausverbrauch abgezogen. Bei der Gesamtenergie der PV addiert auch beide Werte korrekt zusammen.

andig commented on templates/definition/meter/deye-hybrid-hp3.yaml:
```diff
@@ -30,6 +30,10 @@ params:
   - name: maxacpower
+  - name: includegenport
+    type: bool
+    default: false
```
```suggestion
```

Johnny1206 replied:
wieso willst du den default draußen haben? hätte es hier eigentlich für sinnvoll gehalten.

andig replied:
weil false der default von bool ist. Das brauchts einfach nicht

andig commented on templates/definition/meter/deye-hybrid-hp3.yaml:
```diff
@@ -30,6 +30,10 @@ params:
+  - name: includegenport
+    type: bool
+    default: false
+    advanced: true
```
Der Parameter braucht Hilfetext/Description

Johnny1206 replied:
Bitte kurze Info wo ich das eintragen soll. Oder einfach als Code-Comment dahinter?

premultiply replied:
`description:`
Siehe andere Templates

Johnny1206 committed:
Update templates/definition/meter/deye-hybrid-hp3.yaml

Co-authored-by: andig <cpuidle@gmail.com>

Johnny1206 committed:
added description to the "includegenport" parameter

Johnny1206 commented:
@andig Hab Beschreibung hinzugefügt. passt das so?

premultiply committed:
simplify
